File: examples/perl/send_sms.pl
#!/usr/bin/env perl

use strict;
use warnings;
use Encode qw(encode);
use HTTP::Tiny;

if (@ARGV < 4) {
    print STDERR "Usage: perl examples/perl/send_sms.pl BASE_URL TOKEN MOBILE MESSAGE [MESSAGE...]\n";
    print STDERR "Example: perl examples/perl/send_sms.pl http://sms.example.net/sms-gateway \$TOKEN +447700900000 Hello from Perl\n";
    exit 2;
}

my ($base_url, $token, $mobile, @message_parts) = @ARGV;
$base_url =~ s{/+\z}{};
my $message = join ' ', @message_parts;
my $send_url = $base_url . '/send/' . encode_path_segment($mobile);

my $response = HTTP::Tiny->new->post(
    $send_url,
    {
        headers => {
            'X-SMS-Gateway-Token' => $token,
            'Content-Type' => 'text/plain; charset=utf-8',
        },
        content => $message,
    }
);

print $response->{content}, "\n" if length $response->{content};

if (!$response->{success}) {
    print STDERR "Gateway returned HTTP $response->{status}\n";
    exit 1;
}

sub encode_path_segment {
    my ($value) = @_;
    return join '', map {
        /[A-Za-z0-9._~-]/ ? $_ : sprintf '%%%02X', ord $_
    } split //, encode('UTF-8', $value);
}

File: examples/send_sms.py
#!/usr/bin/env python3

import sys
import urllib.error
import urllib.parse
import urllib.request


def main() -> int:
    if len(sys.argv) < 5:
        print(
            "Usage: python send_sms.py BASE_URL TOKEN MOBILE MESSAGE [MESSAGE...]",
            file=sys.stderr,
        )
        print(
            "Example: python send_sms.py http://sms.example.net/sms-gateway "
            "$TOKEN +447700900000 Hello from Python",
            file=sys.stderr,
        )
        return 2

    base_url = sys.argv[1].rstrip("/")
    token = sys.argv[2]
    mobile = sys.argv[3]
    message = " ".join(sys.argv[4:])

    send_url = f"{base_url}/send/{urllib.parse.quote(mobile, safe='')}"
    request = urllib.request.Request(
        send_url,
        data=message.encode("utf-8"),
        method="POST",
        headers={
            "X-SMS-Gateway-Token": token,
            "Content-Type": "text/plain; charset=utf-8",
        },
    )

    try:
        with urllib.request.urlopen(request) as response:
            print(response.read().decode("utf-8"))
            return 0
    except urllib.error.HTTPError as error:
        print(error.read().decode("utf-8"), file=sys.stderr)
        print(f"Gateway returned HTTP {error.code}", file=sys.stderr)
        return 1
    except urllib.error.URLError as error:
        print(f"Request failed: {error.reason}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
